### The purpose of this work is to analyze textual log data from an online chat forum related to the
### Anonymous hacktivist group.

### IRC is an early protocol for instant messaging developed in the early years of the Internet. The openness and ability to remain anonymous has made IRC a popular channel for hacker networks to collaborate and share ideas.

#### The collection and analysis of these chats is a form of cyber-threat intelligence. The analysis of these chats and other dark web data sources enable proactive defense against attacks. 

In [71]:
import re
import collections
import pandas as pd

In [2]:
fp=open('hackers.log','r',encoding='latin-1')

In [3]:
data=fp.read()

In [224]:
data[:3000]

'--- Log opened Tue Sep 20 00:01:49 2016\n00:01 -!- Guest40341 [AndChat2541@AN-pl0gl1.8e2d.64f9.r226rd.IP] has quit [Quit: Bye]\n00:11 -!- peejr [peeejr@AN-sru.3ib.ec0efc.IP] has joined #hackers\n00:14 -!- Gilgamesh [Gilgamesh@AN-nkf.mv0.se355c.IP] has joined #hackers\n00:15 -!- _CyBruh_ [-Cybruh@AN-gm6.oj9.rj1tv4.IP] has quit [Quit: Leaving]\n00:20 -!- peejr [peeejr@AN-sru.3ib.ec0efc.IP] has quit [Quit: Leaving]\n00:25 < ice231> anyone good with exploiting cisco asa with extrabacon?\n00:27 < ice231> we need help with an op but were stuck at this one part\n00:27 -!- Bobseviltwin [steven@stupid.hunkey.monkey] has quit [Ping timeout: 121 seconds]\n00:30 -!- Gilgamesh [Gilgamesh@AN-nkf.mv0.se355c.IP] has quit [Quit: Leaving]\n00:30 -!- peejr [peeejr@AN-sru.3ib.ec0efc.IP] has joined #hackers\n00:34 -!- peejr [peeejr@AN-sru.3ib.ec0efc.IP] has quit [Quit: Leaving]\n00:34 -!- peejr [peeejr@AN-sru.3ib.ec0efc.IP] has joined #hackers\n00:35 -!- Anonymous5 [Anonymous5@AN-tu4.e85.r2ddjo.IP] has jo

### Now going to do some analysis which is useful to get an overall idea of this chat logs of hactivist group

#### We can see in the logs that all the user ip's are in square brackets and there is a pattern that @ will appear in the user ip

#### Extracting all the diffrent user ip's from the chat log using regular expression

In [225]:
clean_1=re.findall(r'\[.+@.+?\]',data)

In [230]:
total_users=set(clean_1)

In [231]:
len(total_users)

34896

#### There are total 34896 diffrent users ip we have in the log

#### Q.How many times log opened and what is the first and last date?

In [16]:
log_open=data.split('--- Log opened')

In [17]:
len(log_open)

559

#### Log has opened total 559 times

In [10]:
clean_1

['[AndChat2541@AN-pl0gl1.8e2d.64f9.r226rd.IP]',
 '[peeejr@AN-sru.3ib.ec0efc.IP]',
 '[Gilgamesh@AN-nkf.mv0.se355c.IP]',
 '[-Cybruh@AN-gm6.oj9.rj1tv4.IP]',
 '[peeejr@AN-sru.3ib.ec0efc.IP]',
 '[steven@stupid.hunkey.monkey]',
 '[Gilgamesh@AN-nkf.mv0.se355c.IP]',
 '[peeejr@AN-sru.3ib.ec0efc.IP]',
 '[peeejr@AN-sru.3ib.ec0efc.IP]',
 '[peeejr@AN-sru.3ib.ec0efc.IP]',
 '[Anonymous5@AN-tu4.e85.r2ddjo.IP]',
 '[resistant@AN-kjb.6km.8lhfko.IP]',
 '[peeejr@AN-sru.3ib.ec0efc.IP]',
 '[peeejr@AN-sru.3ib.ec0efc.IP]',
 '[resistant@AN-kjb.6km.8lhfko.IP]',
 '[peeejr@AN-sru.3ib.ec0efc.IP]',
 '[anon@technoalchemy.com]',
 '[Anonymous5@AN-tu4.e85.r2ddjo.IP]',
 '[ice231@AN-8ak.iq1.hl67p8.IP]',
 '[lulzface@k.e.d.d.y]',
 '[ice231@AN-6ks.fag.lh60mn.IP]',
 '[S0ldierF0re@AN-5ms.4a0.9sk9lp.IP]',
 '[webirc@AN-o8o.sb7.2n2b0e.IP]',
 '[lulz@drone.incoming.pew.pews]',
 '[androirc@AN-l8e.7dp.8hdu3q.IP]',
 '[Unknown@AN-6mk.97v.o0pdt6.IP]',
 '[prudent@AN-r5g.h92.dq7njm.IP]',
 '[prudent@AN-r5g.h92.dq7njm.IP]',
 '[androirc@AN-l

In [11]:
re.findall(r"\[Hex@Quantum.Time].+\n\d\d:\d\d",data)[-1]

'[Hex@Quantum.Time] has joined #hackers\n18:44'

In [12]:
total_users

{'[mib@AN-2do.kd3.79ces3.IP]',
 '[SteppenWolf@AN-hvn.27j.1b8mr0.IP]',
 '[devbaka@AN-a8o.jih.hepm0a.IP]',
 '[psouth@AN-5ndss7.863l.75fn.tn5i6g.IP]',
 '[reverse@AN-jk0.tn1.m68a4i.IP]',
 '[Zer0Z@AN-nlqv4e.hc5h.r1c4.puh5an.IP]',
 '[_dwong879@AN-hlk.cts.51tv80.IP]',
 '[goaway@kek.the.new.years]',
 '[IBLIS@AN-1jm.6be.7s4e54.IP]',
 '[hera@AN-8t0epa.5lup.ab70.6hve8k.IP]',
 '[webirc@AN-0lf.7lq.lkc13k.IP]',
 '[Android@AN-qrj.t5u.v5jg98.IP]',
 '[dxd@Dark.Side]',
 '[Eschiclers@AN-b9j.jt5.edu4ee.IP]',
 '[Dachshund@AN-e2m.b33.hin9nu.IP]',
 '[nil33t@AN-oal.r3g.5p1guh.IP]',
 '[anon_lulz@AN-00o.ihv.pvce23.IP]',
 '[spider876@AN-dsu.g6c.hic3gp.IP]',
 '[webirc@AN-r19.kka.ucf2t7.IP]',
 '[bitch@AN-kp2.ha7.l989to.IP]',
 '[resistant@AN-fh5.ubf.u16ro2.IP]',
 '[fooman@AN-b8n.7o5.iv3f18.IP]',
 '[Unknown@AN-1qk.m3r.8umi4t.IP]',
 '[resistant@AN-80u.iea.vakjoj.IP]',
 '[NoMore@AN-jt6.675.tpavhn.IP]',
 '[itachi@AN-q25vt1.d648.sfhq.07e1cf.IP]',
 '[blackbaron@AN-u1b.5kn.8n5t1i.IP]',
 '[moonbeam@AN-rsr.gch.gs1tgt.IP]',


In [13]:
frequency=dict(collections.Counter(clean_1))

In [14]:
m=max(frequency.values())

In [15]:
for i,j in frequency.items():
    if j==m:
        print(i)

[Hex@Quantum.Time]


#### User "[Hex@Quantum.Time]" Has Logged in maximum time

#### Extract all the urls that have been shared in the chats Day wise

In [117]:
urls=[]
url_count=[]
for i in log_open[1:]:
    url=re.findall(r"\bhttp\S+\b",i)
    urls.append(url)
    url_count.append(len(url))

In [20]:
log_open[:10]

['',
 ' Tue Sep 20 00:01:49 2016\n00:01 -!- Guest40341 [AndChat2541@AN-pl0gl1.8e2d.64f9.r226rd.IP] has quit [Quit: Bye]\n00:11 -!- peejr [peeejr@AN-sru.3ib.ec0efc.IP] has joined #hackers\n00:14 -!- Gilgamesh [Gilgamesh@AN-nkf.mv0.se355c.IP] has joined #hackers\n00:15 -!- _CyBruh_ [-Cybruh@AN-gm6.oj9.rj1tv4.IP] has quit [Quit: Leaving]\n00:20 -!- peejr [peeejr@AN-sru.3ib.ec0efc.IP] has quit [Quit: Leaving]\n00:25 < ice231> anyone good with exploiting cisco asa with extrabacon?\n00:27 < ice231> we need help with an op but were stuck at this one part\n00:27 -!- Bobseviltwin [steven@stupid.hunkey.monkey] has quit [Ping timeout: 121 seconds]\n00:30 -!- Gilgamesh [Gilgamesh@AN-nkf.mv0.se355c.IP] has quit [Quit: Leaving]\n00:30 -!- peejr [peeejr@AN-sru.3ib.ec0efc.IP] has joined #hackers\n00:34 -!- peejr [peeejr@AN-sru.3ib.ec0efc.IP] has quit [Quit: Leaving]\n00:34 -!- peejr [peeejr@AN-sru.3ib.ec0efc.IP] has joined #hackers\n00:35 -!- Anonymous5 [Anonymous5@AN-tu4.e85.r2ddjo.IP] has joined #ha

In [58]:
l=[]
for i in log_open:
    l.append(len(set(re.findall(r"\[.+@.+?]",i))))

In [59]:
index_=l.index(max(l))

In [60]:
max_activity_day=log_open[index_][:20]

In [61]:
max_activity_day

' Mon Oct 17 04:28:44'

### On  'Mon Oct 17 04:28:44' there was maximum traffic

In [62]:
data[:1000]

'--- Log opened Tue Sep 20 00:01:49 2016\n00:01 -!- Guest40341 [AndChat2541@AN-pl0gl1.8e2d.64f9.r226rd.IP] has quit [Quit: Bye]\n00:11 -!- peejr [peeejr@AN-sru.3ib.ec0efc.IP] has joined #hackers\n00:14 -!- Gilgamesh [Gilgamesh@AN-nkf.mv0.se355c.IP] has joined #hackers\n00:15 -!- _CyBruh_ [-Cybruh@AN-gm6.oj9.rj1tv4.IP] has quit [Quit: Leaving]\n00:20 -!- peejr [peeejr@AN-sru.3ib.ec0efc.IP] has quit [Quit: Leaving]\n00:25 < ice231> anyone good with exploiting cisco asa with extrabacon?\n00:27 < ice231> we need help with an op but were stuck at this one part\n00:27 -!- Bobseviltwin [steven@stupid.hunkey.monkey] has quit [Ping timeout: 121 seconds]\n00:30 -!- Gilgamesh [Gilgamesh@AN-nkf.mv0.se355c.IP] has quit [Quit: Leaving]\n00:30 -!- peejr [peeejr@AN-sru.3ib.ec0efc.IP] has joined #hackers\n00:34 -!- peejr [peeejr@AN-sru.3ib.ec0efc.IP] has quit [Quit: Leaving]\n00:34 -!- peejr [peeejr@AN-sru.3ib.ec0efc.IP] has joined #hackers\n00:35 -!- Anonymous5 [Anonymous5@AN-tu4.e85.r2ddjo.IP] has jo

#### Now after doing this analysis making a data frame with all the information extracted

In [92]:
open_dates=re.findall(r"--- Log opened.+\n\d\d:\d\d",data)

In [93]:
logs_info=pd.DataFrame(open_dates,columns=['Dates'])

In [98]:
logs_info['opening_Date']=logs_info['Dates'].apply(lambda x:x.split('opened')[-1].split('\n')[0])

In [99]:
logs_info['opening_time']=logs_info['Dates'].apply(lambda x:x.split('opened')[-1].split('\n')[-1])

In [103]:
logs_info.drop(['Dates'],axis=1,inplace=True)

In [104]:
logs_info

opening_Date opening_time
0     Tue Sep 20 00:01:49 2016        00:01
1     Thu Sep 22 23:57:59 2016        23:57
2     Mon Sep 26 14:48:05 2016        14:48
3     Wed Sep 28 22:45:15 2016        22:45
4     Fri Sep 30 17:53:45 2016        17:53
..                         ...          ...
553   Thu Apr 26 21:22:07 2018        21:22
554   Fri Apr 27 21:22:08 2018        21:22
555   Sat Apr 28 21:22:09 2018        21:22
556   Sun Apr 29 21:22:09 2018        21:22
557   Mon Apr 30 21:22:08 2018        21:22

[558 rows x 2 columns]

In [118]:
logs_info['users_joined']=l[1:]
logs_info['urls_shared']=urls
logs_info['url_count']=url_count

In [119]:
logs_info

opening_Date opening_time  users_joined  \
0     Tue Sep 20 00:01:49 2016        00:01           414   
1     Thu Sep 22 23:57:59 2016        23:57           465   
2     Mon Sep 26 14:48:05 2016        14:48           320   
3     Wed Sep 28 22:45:15 2016        22:45           257   
4     Fri Sep 30 17:53:45 2016        17:53           384   
..                         ...          ...           ...   
553   Thu Apr 26 21:22:07 2018        21:22            95   
554   Fri Apr 27 21:22:08 2018        21:22            89   
555   Sat Apr 28 21:22:09 2018        21:22            94   
556   Sun Apr 29 21:22:09 2018        21:22            96   
557   Mon Apr 30 21:22:08 2018        21:22            89   

                                           urls_shared  url_count  
0    [http://i.imgur.com/PoCjYqQ.png, https://vid.p...         39  
1    [http://www.androirc.com, http://www.androirc....         77  
2    [http://i.imgur.com/rWAnqP3.jpg, https://twitt...         32  
3    [https://sourceforge.net/projects/qaircrack-ng...         26  
4    [http://www.androirc.com, http://znc.in, https...         69  
..                                                 ...        ...  
553                                        [httpflood]          1  
554  [https://www.youtube.com/watch?v=nfBxNDjlBo4, ...          9  
555  [https://imgur.com/a/Qm7VrMt, https://www.yout...         17  
556  [https://github.com/deptofdefense/gr-uaslink, ...          9  
557      [https://www.youtube.com/watch?v=IBvf7KUEZ78]          1  

[558 rows x 5 columns]

### Finding All the messages in the log

In [186]:
log_messages_info=pd.DataFrame([],columns=['message','user_name','msg_timing','Date'])

In [207]:
for i in log_open[1:]:
    log_msg=pd.DataFrame(re.findall(r"<.+>.*\n\d\d:\d\d",i),columns=['messages'])
    log_msg['msg_timing']=log_msg['messages'].apply(lambda x:x.split('\n')[-1])
    log_msg['message']=log_msg['messages'].apply(lambda x : x.split('>')[-1].split('\n')[0])
    log_msg['user_name']=log_msg['messages'].apply(lambda x : x.split('>')[0])
    log_msg.drop(['messages'],axis=1,inplace=True)
    log_msg['Date']=i[:25]
    log_messages_info=pd.concat([log_messages_info,log_msg])

### Now finally we have two data frames "log_messages_info" and "logs_info" which contains all the information we extracted through the analysis above

In [220]:
log_messages_info.head(5)

message      user_name  \
0   anyone good with exploiting cisco asa with ex...       < ice231   
1   we need help with an op but were stuck at thi...       < ice231   
2                                             hello?  < HeavenGuard   
3                                                 hi       <+nemecy   
4                                                 hi       < ice231   

  msg_timing                       Date  
0      00:27   Tue Sep 20 00:01:49 2016  
1      00:27   Tue Sep 20 00:01:49 2016  
2      01:17   Tue Sep 20 00:01:49 2016  
3      01:22   Tue Sep 20 00:01:49 2016  
4      01:25   Tue Sep 20 00:01:49 2016

In [221]:
logs_info.head(5)

opening_Date opening_time  users_joined  \
0   Tue Sep 20 00:01:49 2016        00:01           414   
1   Thu Sep 22 23:57:59 2016        23:57           465   
2   Mon Sep 26 14:48:05 2016        14:48           320   
3   Wed Sep 28 22:45:15 2016        22:45           257   
4   Fri Sep 30 17:53:45 2016        17:53           384   

                                         urls_shared  url_count  
0  [http://i.imgur.com/PoCjYqQ.png, https://vid.p...         39  
1  [http://www.androirc.com, http://www.androirc....         77  
2  [http://i.imgur.com/rWAnqP3.jpg, https://twitt...         32  
3  [https://sourceforge.net/projects/qaircrack-ng...         26  
4  [http://www.androirc.com, http://znc.in, https...         69